### PR TITLE
DBus-API: implement transient socket unit

### PR DESCRIPTION
**Pull request by yuwata**
This implement transient socket unit. On top of that, now `systemd-run` can create socket and path transient units. Also, this splits `bus_append_unit_property_assignment()` into functions corresponds to unit types.

### CODE REVIEW

**Comment by yuwata** [comment]
Several commits which simplify `bus_append_unit_property_assignment()` are added. Please take a look.

**Comment by yuwata** [comment]
Updated. This may be too large. Sorry...

**Comment by poettering** [comment]
Quit note, I prepped a coccinelle script for fixing the strv_length() things automatically in PR #7729. It finds the same spots. I didn't include the fixes in that PR though, so not to conflict with this PR in this regard.

File: src/core/dbus-service.c [hunk 1]
```diff
@@ -88,6 +88,121 @@ const sd_bus_vtable bus_service_vtable[] = {
         SD_BUS_VTABLE_END
 };
 
+int bus_exec_command_set_transient_property(
```
Comment (poettering) [inline src/core/dbus-service.c]: this new function should probably move to dbus-execute.c

File: src/core/dbus-socket.c [hunk 2]
```diff
@@ -162,6 +171,473 @@ const sd_bus_vtable bus_socket_vtable[] = {
...
+
+        } else if (streq(name, "SocketProtocol")) {
+                int32_t i;
+
+                r = sd_bus_message_read(message, "i", &i);
+                if (r < 0)
+                        return r;
+
+                if (!IN_SET(i, IPPROTO_UDPLITE, IPPROTO_SCTP))
+                        return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "Invalide %s: %i", name, i);
```
Comment (poettering) [inline src/core/dbus-socket.c]: Invalide → Invalid

File: src/core/dbus-socket.c [hunk 3]
```diff
@@ -162,6 +171,473 @@ const sd_bus_vtable bus_socket_vtable[] = {
...
+
+                while ((r = sd_bus_message_read(message, "(ss)", &t, &a)) > 0) {
+                        _cleanup_free_ SocketPort *p = NULL;
+
+                        p = new0(SocketPort, 1);
+                        if (!p)
+                                return log_oom();
+
+                        if (STR_IN_SET(t, "Stream", "Datagram", "SequentialPacket", "Netlink"))
+                              p->type = SOCKET_SOCKET;
```
Comment (poettering) [inline src/core/dbus-socket.c]: indentation is off

File: src/core/dbus-socket.c [hunk 4]
```diff
@@ -162,6 +171,473 @@ const sd_bus_vtable bus_socket_vtable[] = {
...
+                        else if (streq(t, "Special"))
+                                p->type = SOCKET_SPECIAL;
+                        else if (streq(t, "MessageQueue"))
+                                p->type = SOCKET_MQUEUE;
+                        else if (streq(t, "FIFO"))
+                                p->type = SOCKET_FIFO;
+                        else if (streq(t, "USBFunction"))
+                                p->type = SOCKET_USB_FUNCTION;
```
Comment (poettering) [inline src/core/dbus-socket.c]: i figure we should add a translation table for this? i.e. the usual DEFINE_STRING_TABLE_LOOKUP() stuff?

File: src/core/dbus-socket.c [hunk 5]
```diff
@@ -162,6 +171,473 @@ const sd_bus_vtable bus_socket_vtable[] = {
...
+                                        return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "Invalid netlink address: %s", a);
+
+                        } else {
+                                r = socket_address_parse(&p->address, a);
+                                if (r < 0)
+                                        return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "Invalid address: %s", a);
+
+                                if (socket_address_family(&p->address) != AF_LOCAL && p->address.type == SOCK_SEQPACKET) {
+                                        return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "Address family not supported: %s", a);
+                                }
```
Comment (poettering) [inline src/core/dbus-socket.c]: single-line if blocks no {}...

Comment (poettering) [reply src/core/dbus-socket.c]: this check needs to be moved a bit down, after the `p->address.type` initialization

File: src/core/dbus-socket.c [hunk 6]
```diff
@@ -162,6 +171,473 @@ const sd_bus_vtable bus_socket_vtable[] = {
...
+                        empty = false;
+
+                        if (!UNIT_WRITE_FLAGS_NOOP(flags)) {
+                                SocketPort *tail;
+
+                                if (s->ports) {
+                                        LIST_FIND_TAIL(port, s->ports, tail);
+                                        LIST_INSERT_AFTER(port, s->ports, tail, p);
+                                } else
+                                        LIST_PREPEND(port, s->ports, p);
```
Comment (poettering) [inline src/core/dbus-socket.c]: hmm, i am pretty sure LIST_FIND_TAIL+LIST_INSERT_AFTER does the right thing also when the list is empty, no?

File: src/shared/bus-unit-util.c [hunk 7]
```diff
@@ -70,435 +71,420 @@ int bus_parse_unit_info(sd_bus_message *message, UnitInfo *u) {
                         &u->job_path);
 }
 
-static int bus_append_ip_address_access(sd_bus_message *m, int family, const union in_addr_union *prefix, unsigned char prefixlen) {
-        int r;
+#define DEFINE_BUS_APPEND_PARSE_PTR(sd_type, cast_type, type, parse_func)                            \
```
Comment (poettering) [inline src/shared/bus-unit-util.c]: sd_type → bus_type? after all this is the type system of DBus, not of systemd

File: src/basic/socket-util.c [hunk 8]
```diff
@@ -870,6 +870,13 @@ static const char* const ip_tos_table[] = {
 
 DEFINE_STRING_TABLE_LOOKUP_WITH_FALLBACK(ip_tos, int, 0xff);
 
+static const char* const socket_protocol_table[] = {
+        [IPPROTO_UDPLITE] = "udplite",
+        [IPPROTO_SCTP] = "sctp",
+};
```
Comment (poettering) [inline src/basic/socket-util.c]: hmm, when we add this as proper api to socket-util.c we should probably make this table complete with all known IPPROTO_xyz definitions

File: src/core/dbus-cgroup.c [hunk 9]
```diff
@@ -714,6 +717,9 @@ int bus_cgroup_set_property(
 
                 while ((r = sd_bus_message_read(message, "(st)", &path, &weight)) > 0) {
 
+                        if (!path_startswith(path, "/dev"))
```
Comment (poettering) [inline src/core/dbus-cgroup.c]: hmm, we should allow path_startswith(path, "/run/systemd/inaccessible/") here too I figure, as we create 0:0 device nodes there and those should be includable in a device policy

**Comment by yuwata** [comment]
@poettering Thank you for many comments. I've force-pushed updated version. Please take a look.
Main changes are,
- add `socket_protocol_from_name()` and `socket_protocol_to_name()` which covers all protocol name.
- CGroup's IO related settings e.g. `IODeviceWeight=` now accept device node files in /run/systemd/inaccessible/.

**Comment by poettering** [comment]
excellent! thanks! lgtm!

**Comment by poettering** [comment]
CI failure: `KVM: entry failed, hardware error 0x0`

**Pull request merged**